Wangscape/image/Image.cpp

```diff
@@ -30,3 +63,254 @@ ImageGrey imageToGreyscale(const ImageColour & image)
+    const auto nz_max = arma::max(nonzero_coords, 1).eval();
+    const auto nz_size = (nz_max - nz_min + 1).eval();
+    return sf::IntRect(nz_min(1), nz_min(0), nz_size(1), nz_size(0));
+}
+
+sf::IntRect boxUnion(sf::IntRect a, sf::IntRect b)
+{
+    arma::imat origins = {{a.left, a.top}, {b.left, b.top}};
+    arma::imat sizes = {{a.width, a.height}, {b.width, b.height}};
+    auto ends = (origins + sizes - 1).eval();
+    auto origin_min = arma::min(origins, 0).eval();
+    auto end_max = arma::max(ends, 0).eval();
+    auto size_max = (end_max - origin_min + 1).eval();
+    return sf::IntRect(origin_min(0), origin_min(1),
+                       size_max(0), size_max(1));
+}
+
+ImageGrey32 distances(const ImageGrey & traversable, const ImageGrey & targets, bool use_diagonals)
+{
+    if (!isBinary(traversable))
+        throw std::runtime_error("distances() requires binary traversable mask");
+    if (!isBinary(targets))
+        throw std::runtime_error("distances() requires binary targets mask");
+    if(!isNonzero(traversable))
+        throw std::runtime_error("distances() requires nonzero traversable mask");
+    if (!isNonzero(targets))
+        throw std::runtime_error("distances() requires nonzero targets mask");
+    if(!isConnected(traversable))
+        throw std::runtime_error("distances() requires connected traversable mask");
+    const auto shape = makeUVec(arma::size(traversable));
+    ImageGrey32 distance_field(arma::size(traversable));
+    const arma::u32 infinity = std::numeric_limits<arma::u32>::max();
+    distance_field.fill(infinity);
+    const auto target_coordinates = arma::ind2sub(arma::size(targets), arma::find(targets)).eval();
+    std::queue<IVec> border;
+    const auto& steps = use_diagonals ? ORTHODIAGONAL_STEPS : ORTHOGONAL_STEPS;
+    const auto in_bounds = [&shape](IVec point)
+    {
+        if (point.x() < 0 || point.y() < 0)
+            return false;
+        if ((unsigned int)point.x() >= shape.x() || (unsigned int)point.y() >= shape.y())
+            return false;
+        return true;
+    };
+    for (unsigned int i = 0; i < target_coordinates.n_cols; i++)
+    {
+        IVec target{int(target_coordinates(1, i)), int(target_coordinates(0, i))};
+        distance_field(target.y(), target.x()) = 0;
+        border.push(target);
+    }
+    while (!border.empty())
+    {
+        IVec point = border.front();
+        border.pop();
+        arma::u32 distance = distance_field(point.y(), point.x()) + 1;
+        for (const auto step : steps)
+        {
+            IVec new_point = point + step;
+            if (!in_bounds(new_point))
+                continue;
+            if (!traversable(new_point.y(), new_point.x()))
+                continue;
+            if (distance_field(new_point.y(), new_point.x()) == infinity)
+            {
+                border.push(new_point);
+                distance_field(new_point.y(), new_point.x()) = distance;
+            }
+        }
```

Comment:
I'd consider extracting this neighborhood-mapping part... but I'm not sure how (what arguments it should accept). I suppose that it will be repeated one by one in future code.

Reply:
Can you clarify this?

Reply:
I mean defining a function to traverse through matrix and map elements with a given function and neighborhood. I'm not sure if:
* it's worth an effor
* is there an easy way to do so.

So I'm only suggesting it.

I've tried to do it in one of my projects and it works this way:
* I've defined a function, which traverses a given matrix `A`, executing a given function `f` on its elements:

https://github.com/hryniuk/image-processing/blob/19f2fa10441a586657e2f5f44066bb8052116064/op/IndexForEach.hpp#L14-L26

* I've defined an operators that can be passed to the above function, which are calculating a value based on neighborhood (`surr`oundings):

https://github.com/hryniuk/image-processing/blob/19f2fa10441a586657e2f5f44066bb8052116064/op/Mean.hpp#L16-L31

And this inner `for` loop can be generalized as a function that executes a value with a given neighborhood.

**But**:
* for more complex operations it could be harder to read and analyze
* for simple things there is `arma::conv2`

sooo... ¯\\\_(ツ)_/¯.

Reply:
I think it will worthwhile to split `distance()` up and try to unify its implementation with `isConnected()`. Some of the ideas in #37 also require similar operations. However, this is fundamentally a graph algorithm (breadth first search) - the order of processing is determined by the numerical values of the input arrays, not their size. Trying to iterate over the whole matrix row-by-row would obscure what the algorithm does, rather than clarify it.

Reply:
I've extracted the BFS component of `distances` and `isConnected` into a function, but it uses four different callbacks. Perhaps it would be better to make it a functor with 4 protected pure virtual functions.